paragraph/graph/graph.cc

```diff
@@ -415,6 +420,8 @@ absl::Status Graph::ValidateCommon() const {
       instruction_ids.insert(instruction->GetId());
     }
   }
+  // Check graph is fully connected
+  RETURN_IF_ERROR(IsConnected());
```

Comment:
Wouldn't it make more sense to fail before doing the work?

Reply:
Amount of work is comparable for `IsConnected()` and previous checks. Previous checks figure out if it's even safe to crawl the graph, i.e. there is root, instructions IDs are unique, subroutines point to the right graph. Once that's clear, we can check that instructions in every subroutine are connected